Low Fuel Motorsport count down speech synthesizer

iFuSiiOnzZ opened the pull request:
## Functionality
There are 5 timers:
   - Number of minutes until race stars
   - 5 minutes
   - 3 minutes
   - 1 minute
   - Race started (time 0)

If user withdraw from the race the timers will be removed and a "Race cancelled" will be added instead.
Once that the race has started the timer identifiers are removed.

## Infrastructure
There are 2 main files. 

- One for data structure which contains information about the task to run concurrent list implementation where the elements are sorted by the execution date.
- The other file is the logic of the timer task class. As elements in the queue are always ordered, only checks if the first element has to executed, otherwise it will sleep for a certain amount of milliseconds. When a task has to be executed it created C# task that runs it asynchronously.

## Important
- Not tested (test requires time)
- This could lead to many threads creation if is have used

iFuSiiOnzZ committed:
Add infrastructure to execute a task at a given point in time

iFuSiiOnzZ committed:
Add LowFuelMotorsport count down speech synthesizer

There are 4 timer:
   5 minutes
   3 minutes
   1 minute
   Race started (time 0)

If user withdraw from the race the timers will be removed and a "Race canceled" will be added instead.
Once that the race has started the timer identifiers are removed.

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

iFuSiiOnzZ committed:
Improve task timer executor infrastructure

iFuSiiOnzZ committed:
Improve LFM speech synthesizer

RiddleTime commented on Race Element.Core/Jobs/Timer/DataStructures.cs:
```diff
@@ -0,0 +1,114 @@
+    /// <returns>True if element peeked, else otherwise.</returns>
+    public bool TryPeekFront(out TimerData data)
+    {
+        lock (_list)
```
when using `lock(..)` use the System.Threading.Lock as an object to lock on 
https://learn.microsoft.com/en-us/dotnet/api/system.threading.lock?view=net-9.0#remarks

iFuSiiOnzZ committed:
Rename function from "AddTimer" to "Add"

iFuSiiOnzZ committed:
Use System.Threading.Lock

iFuSiiOnzZ committed:
Update doxygen

iFuSiiOnzZ committed:
Fix first speech message

RiddleTime commented on Race_Element.HUD.ACC/Overlays/Pitwall/OverlayLowFuelMotorsport/LowFuelMotorsportSpeechSynthesizer.cs:
```diff
@@ -0,0 +1,32 @@
+
+internal class LowFuelMotorsportSpeechSynthesizer : IJob
+{
+    private readonly SpeechSynthesizer _synth;
```
Where is the `_synth?.Dispose()` happening?

iFuSiiOnzZ replied:
Refactoring the `Run()` function to call `Speak` instead of `SpeakAsync()` in order to call `Dispose()`.

RiddleTime commented on Race_Element.HUD.ACC/Overlays/Pitwall/OverlayLowFuelMotorsport/LowFuelMotorsportSpeechSynthesizer.cs:
```diff
@@ -0,0 +1,32 @@
+
+internal class LowFuelMotorsportSpeechSynthesizer : IJob
+{
+    private readonly SpeechSynthesizer _synth;
```
How does the SpeechSynthesizer fare when 2 instances of it are initialized and being used for speech?

(I.E. have 2 huds running that both contain a seperate instance of the speech synthesizer )

iFuSiiOnzZ replied:
It mixes the streams, so you will have both sounds overlapping.

RiddleTime commented on Race_Element.HUD.ACC/Overlays/Pitwall/OverlayLowFuelMotorsport/LowFuelMotorsportSpeechSynthesizer.cs:
```diff
@@ -0,0 +1,32 @@
+    private readonly SpeechSynthesizer _synth;
+    private readonly String _message;
+
+    private LowFuelMotorsportOverlay _overlay;
```
if only initialized once, readonly can be applied as it's being initialized in the constructor

iFuSiiOnzZ replied:
Done

RiddleTime commented on Race_Element/App.xaml.cs:
```diff
@@ -220,6 +220,7 @@ private void HandleCustomUriScheme(string arg)
 
     private void App_Exit(object sender, ExitEventArgs e)
     {
+        TaskTimerExecutor.Instance().Dispose();
```
Yes.

RiddleTime replied:
This is a good point to refactor all existing long running "tasks" into. (regard this as a note to self but also yes I like it!)

iFuSiiOnzZ committed:
Add `sealed` keyword to classes

iFuSiiOnzZ committed:
Refactoring the `Run()` function to call `Speak` instead of `SpeakAsync()` in order to call `Dispose()`.

iFuSiiOnzZ committed:
Rename wait event

iFuSiiOnzZ committed:
Simplify logic for speech synthesizer message

iFuSiiOnzZ committed:
Rename class from "TaskTimerExecutor" to "JobTimerExecutor"

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

RiddleTime committed:
Seal reference property

additionaly use a primary constructor and set the property _typeReference to readonly

RiddleTime committed:
small refactor

- auto formatted
- add readonly to _list
- rename Callback to Job

RiddleTime committed:
refactor

- add nullref exception checks
- remove extra function call
- add readonly where possible and applicable

RiddleTime committed:
add final readonly modifier

iFuSiiOnzZ committed:
Rename function from "RemoveTimer" to "Remove"

RiddleTime committed:
job time executor reactored

- removed data structures class for job timer and placed all required classes internally in JobTimerExecutor class
- replaced concurrent list timer data with internal c# concurrent collection
- still makes use of event wait handle to continue the worker loop in the executor
- timer jobs are now identified by their job id
- Duplicate job guids are not allowed

RiddleTime committed:
Update JobTimerExecutor.cs

small fix, obviously want to run the worker directly if the upcoming job has no wait time left

RiddleTime committed:
Update JobTimerExecutor.cs

well this is sleeker..

RiddleTime committed:
lfm overlay/speech

job ids are now always collected in the overlay ready to be cancelled if needed

RiddleTime committed:
fix syntax error

RiddleTime committed:
Update JobTimerExecutor.cs

cancel all existing jobs on disposal of job timer executor

RiddleTime committed:
Update JobTimerExecutor.cs

jobs now get removed if they were started but are not running anymore (completed state)

RiddleTime committed:
Update LowFuelMotorsportOverlay.cs

fix speech reminder, added check to see if any jobs existed.

RiddleTime committed:
set ids in private prop for IJob implementations

RiddleTime commented:
I will merge this commit as there are other systems I want to integrate with the job scheduler.
